presto-main/src/main/java/com/facebook/presto/cost/RowNumberStatsRule.java

```diff
@@ -0,0 +1,84 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.cost;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.plan.Patterns;
+import com.facebook.presto.sql.planner.plan.RowNumberNode;
+
+import java.util.Optional;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static java.lang.Double.isNaN;
+import static java.lang.Math.min;
+
+public class RowNumberStatsRule
+        extends SimpleStatsRule<RowNumberNode>
+{
+    private static final Pattern<RowNumberNode> PATTERN = Patterns.rowNumber();
+
+    public RowNumberStatsRule(StatsNormalizer normalizer)
+    {
+        super(normalizer);
+    }
+
+    @Override
+    public Pattern<RowNumberNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Optional<PlanNodeStatsEstimate> doCalculate(RowNumberNode node, StatsProvider statsProvider, Lookup lookup, Session session, TypeProvider types)
+    {
+        PlanNodeStatsEstimate sourceStats = statsProvider.getStats(node.getSource());
+        if (sourceStats.isOutputRowCountUnknown()) {
+            return Optional.of(PlanNodeStatsEstimate.unknown());
+        }
+
+        double partitionCount = 1;
+        for (Symbol groupBySymbol : node.getPartitionBy()) {
+            SymbolStatsEstimate symbolStatistics = sourceStats.getSymbolStatistics(groupBySymbol);
+            int nullRow = (symbolStatistics.getNullsFraction() == 0.0) ? 0 : 1;
+            partitionCount *= symbolStatistics.getDistinctValuesCount() + nullRow;
+        }
+        partitionCount = min(sourceStats.getOutputRowCount(), partitionCount);
+
+        if (isNaN(partitionCount)) {
+            return Optional.of(PlanNodeStatsEstimate.unknown());
+        }
+
+        double sourceRowsCount = sourceStats.getOutputRowCount();
+        double rowsPerPartition = sourceRowsCount / partitionCount;
+        if (node.getMaxRowCountPerPartition().isPresent()) {
+            rowsPerPartition = min(rowsPerPartition, node.getMaxRowCountPerPartition().get());
+        }
+
+        double outputRowsCount = partitionCount * rowsPerPartition;
+        return Optional.of(PlanNodeStatsEstimate.buildFrom(sourceStats)
+                .setOutputRowCount(outputRowsCount)
+                .addSymbolStatistics(node.getRowNumberSymbol(), SymbolStatsEstimate.builder()
+                        .setLowValue(0)
+                        .setHighValue(rowsPerPartition)
```

Comment:
I'm wondering if we need to set the low and high at all. Probably just NDV should be enough (similarly to AssignUniqueId)

Reply:
You can definitely include low value, since you know it.
High value -- you can omit it, since it's not easy to calculate. Leave a code comment explaining why you're doing so.